add ref count to alts shared resource.

yihuazhang opened the pull request:
This PR adds a ref count to ALTS shared resources, and ref (unref) it whenever creating (destroying) alts server/channel credentials. It will resolve ALTS (and google default creds) test flakiness. 

I will prepare another PR later on that removes the TSI thread from gRPC use cases.

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                               FILE SIZE
 ++++++++++++++ GROWING                                 ++++++++++++++
  +0.0%    +332 [None]                                  +9.19Ki  +0.1%
      +0.1%    +364 [Unmapped]                              +9.19Ki  +0.1%
      +6.2%      +8 g_alts_resource                               0  [ = ]
   +48%    +268 src/core/tsi/alts_transport_security.cc    +268   +48%
      [NEW]    +215 grpc_tsi_alts_shared_resource_unref        +215  [NEW]
      [NEW]     +85 grpc_tsi_alts_wait_for_cq_drain             +85  [NEW]
      +108%     +26 [Unmapped]                                  +26  +108%
       +23%     +14 grpc_tsi_alts_init                          +14   +23%
      [NEW]     +12 grpc_tsi_alts_shared_resource_ref           +12  [NEW]

  +0.0%    +600 TOTAL                                   +9.45Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

markdroth commented on src/core/tsi/alts_transport_security.cc:
```diff
@@ -50,6 +50,7 @@ void grpc_tsi_alts_init() {
+  gpr_ref_init(&g_alts_resource.refcount, 0);
 }
 
 void grpc_tsi_alts_shutdown() {
```
If this gets called, then we will shut down the resources even if the refcount hasn't gone to zero.

jiangtaoli2016 replied:
Will this ever happen? Credentials will be destroyed before grpc_shutdown is called.

If it is possible grpc_shutdown is called before refcount > 0, we still did not solve root cause of crash.

markdroth replied:
If that's the case, then shouldn't this code simply assert that `g_alts_resource.cq == nullptr`?

My point here is that we shouldn't have two paths to make this shutdown happen.  If the refcount should always be zero by the time we get here, then there's no reason to ever try to do the shutdown here.

yihuazhang replied:
Another reason for the existence of this chunk of code is that Envoy does will not use ref-count (because they do not use any credential object), so it still relies on `grpc_tsi_alts_shutdown` to destroy channels, cq and so on.

markdroth replied:
I don't think it makes sense to have an API that can be used in both a ref-counted and non-ref-counted fashion.  We should pick one of the two approaches and stick with it.  If this is going to be ref-counted and Envoy is going to use it, then Envoy should be responsible for taking a ref and releasing it when it is no longer needed.

yihuazhang replied:
It seems non-trivial to let Envoy to use ref-count since they directly use `alts_tsi_handshaker`, and it does not make sense to ref (or unref) shared resources in TSI handshakers as they need to be associated with credential objects. Do you have any suggestion Mark? If not, I will discard this PR and directly work on removing TSI threads.

markdroth replied:
Why is it hard for Envoy to do this?  They would just need to take a single ref at startup and unref it at shutdown.

Even if we remove the TSI thread for the gRPC use-case, Envoy will still need its own thread, so I don't think that will solve the problem.

markdroth commented on src/core/tsi/alts_transport_security.cc:
```diff
@@ -63,3 +64,15 @@ void grpc_tsi_alts_shutdown() {
   gpr_mu_destroy(&g_alts_resource.mu);
 }
+
+void grpc_tsi_alts_shared_resource_ref() { gpr_ref(&g_alts_resource.refcount); }
```
What happens if the application creates a channel and then destroys it (thus sending the refcount to 0) and then creates another channel?  We need some way to recreate the shared resources when the second channel is created.

jiangtaoli2016 replied:
Initialization is lazily at https://github.com/grpc/grpc/blob/master/src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc#L333

markdroth replied:
Let me say this a different way: In terms of the API here, it seems odd to me that `grpc_tsi_alts_shared_resource_ref()` and `alts_get_shared_resource()` are different methods.  Shouldn't the API be that there is a single method that is responsible for returning a pointer along with a new ref, which the caller holds until they explicitly unref it?  Otherwise, the API allows weird things like taking a new ref when the previous refcount was 0 and no object actually exists.

yihuazhang replied:
I agree it will be better to have an API that returns a pointer along with the ref. Since in the current ALTS impl, the shared resource is needed at the TSI layer, we need to call `grpc_tsi_alts_shared_resource_ref()` at the credential layer and plumb it all the way down to TSI. I will update the code.

jiangtaoli2016 commented on src/core/tsi/alts_transport_security.cc:
```diff
@@ -63,3 +64,15 @@ void grpc_tsi_alts_shutdown() {
+
+void grpc_tsi_alts_shared_resource_unref() {
+  if (gpr_unref(&g_alts_resource.refcount)) {
+    grpc_completion_queue_shutdown(g_alts_resource.cq);
```
Lines 72-76 and Lines 58-62 can be refactored into a single function to avoid code duplication

jiangtaoli2016 commented on src/core/tsi/alts_transport_security.cc:
```diff
@@ -63,3 +64,15 @@ void grpc_tsi_alts_shutdown() {
+void grpc_tsi_alts_shared_resource_ref() { gpr_ref(&g_alts_resource.refcount); }
+
+void grpc_tsi_alts_shared_resource_unref() {
+  if (gpr_unref(&g_alts_resource.refcount)) {
```
You probably need to get lock before you release the resource.

yihuazhang replied:
Good point. The lock is applied.

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,151      Total (>)      1,961,099

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,884,175      Total (>)     10,883,611

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                               FILE SIZE
 ++++++++++++++ GROWING                                 ++++++++++++++
  +0.0%    +332 [None]                                  +9.19Ki  +0.1%
      +0.1%    +364 [Unmapped]                              +9.19Ki  +0.1%
      +6.2%      +8 g_alts_resource                               0  [ = ]
   +48%    +268 src/core/tsi/alts_transport_security.cc    +268   +48%
      [NEW]    +215 grpc_tsi_alts_shared_resource_unref        +215  [NEW]
      [NEW]     +85 grpc_tsi_alts_wait_for_cq_drain             +85  [NEW]
      +108%     +26 [Unmapped]                                  +26  +108%
       +23%     +14 grpc_tsi_alts_init                          +14   +23%
      [NEW]     +12 grpc_tsi_alts_shared_resource_ref           +12  [NEW]

  +0.0%    +600 TOTAL                                   +9.45Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                           FILE SIZE
 ++++++++++++++ GROWING                                                             ++++++++++++++
  +0.0%    +177 [None]                                                              +19.3Ki  +0.2%
      +0.0%    +209 [Unmapped]                                                          +19.3Ki  +0.2%
      +6.2%      +8 g_alts_resource                                                           0  [ = ]
  +2.0%     +91 src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc                     +91  +2.0%
      [NEW]    +848 alts_tsi_handshaker_create                                             +848  [NEW]
  +6.5%     +36 src/core/tsi/alts_transport_security.cc                                 +36  +6.5%
      [NEW]    +271 grpc_tsi_alts_shared_resource_unref                                    +271  [NEW]
      [NEW]     +28 grpc_tsi_alts_shared_resource_ref                                       +28  [NEW]
       +23%     +14 grpc_tsi_alts_init                                                      +14   +23%
  +7.7%     +32 src/core/lib/security/credentials/alts/alts_credentials.cc              +32  +7.7%
       +11%     +16 grpc_alts_credentials_create_customized                                 +16   +11%
       +11%     +16 grpc_alts_server_credentials_create_customized                          +16   +11%
       +39%      +9 alts_credentials_destruct                                                +9   +39%
       +39%      +9 alts_server_credentials_destruct                                         +9   +39%
  +0.7%     +16 src/core/lib/security/security_connector/alts_security_connector.cc     +16  +0.7%
      +5.0%      +8 alts_server_add_handshakers                                              +8  +5.0%
      +4.8%      +8 alts_channel_add_handshakers                                             +8  +4.8%

  +0.0%    +352 TOTAL                                                               +19.4Ki  +0.2%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                           FILE SIZE
 ++++++++++++++ GROWING                                                             ++++++++++++++
  +0.0%    +177 [None]                                                              +19.3Ki  +0.2%
      +0.0%    +209 [Unmapped]                                                          +19.3Ki  +0.2%
      +6.2%      +8 g_alts_resource                                                           0  [ = ]
  +2.0%     +91 src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc                     +91  +2.0%
      [NEW]    +848 alts_tsi_handshaker_create                                             +848  [NEW]
  +6.5%     +36 src/core/tsi/alts_transport_security.cc                                 +36  +6.5%
      [NEW]    +271 grpc_tsi_alts_shared_resource_unref                                    +271  [NEW]
      [NEW]     +28 grpc_tsi_alts_shared_resource_ref                                       +28  [NEW]
       +23%     +14 grpc_tsi_alts_init                                                      +14   +23%
  +0.7%     +16 src/core/lib/security/security_connector/alts_security_connector.cc     +16  +0.7%
      +5.0%      +8 alts_server_add_handshakers                                              +8  +5.0%
      +4.8%      +8 alts_channel_add_handshakers                                             +8  +4.8%

  +0.0%    +320 TOTAL                                                               +19.4Ki  +0.2%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                           FILE SIZE
 ++++++++++++++ GROWING                                                             ++++++++++++++
  +0.0%    +273 [None]                                                              +19.3Ki  +0.2%
      +0.0%    +305 [Unmapped]                                                          +19.3Ki  +0.2%
      +6.2%      +8 g_alts_resource                                                           0  [ = ]
   +20%    +108 src/core/tsi/alts_transport_security.cc                                +108   +20%
      [NEW]    +279 grpc_tsi_alts_shared_resource_unref                                    +279  [NEW]
      [NEW]     +28 grpc_tsi_alts_shared_resource_ref                                       +28  [NEW]
       +23%     +14 grpc_tsi_alts_init                                                      +14   +23%
       +12%      +3 [Unmapped]                                                               +3   +12%
  +2.0%     +91 src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc                     +91  +2.0%
      [NEW]    +848 alts_tsi_handshaker_create                                             +848  [NEW]
  +0.7%     +16 src/core/lib/security/security_connector/alts_security_connector.cc     +16  +0.7%
      +5.0%      +8 alts_server_add_handshakers                                              +8  +5.0%
      +4.8%      +8 alts_channel_add_handshakers                                             +8  +4.8%

  +0.0%    +488 TOTAL                                                               +19.5Ki  +0.2%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,023      Total (<)      1,961,099

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,883,813      Total (>)     10,883,607

 No significant differences in binary sizes


```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,194      Total (>)      1,961,099

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,883,815      Total (>)     10,883,611

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,194      Total (>)      1,961,099

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,883,812      Total (>)     10,883,607

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                           FILE SIZE
 ++++++++++++++ GROWING                                                             ++++++++++++++
  +0.1%    +385 [None]                                                              +19.3Ki  +0.2%
      +0.1%    +417 [Unmapped]                                                          +19.3Ki  +0.2%
      +6.2%      +8 g_alts_resource                                                           0  [ = ]
   +30%    +164 src/core/tsi/alts_transport_security.cc                                +164   +30%
      [NEW]    +287 grpc_tsi_alts_shared_resource_unref                                    +287  [NEW]
      [NEW]     +52 grpc_tsi_alts_shared_resource_ref                                       +52  [NEW]
       +79%     +19 [Unmapped]                                                              +19   +79%
       +23%     +14 grpc_tsi_alts_init                                                      +14   +23%
      [NEW]      +8 grpc_tsi_alts_shared_resource_get_mu                                     +8  [NEW]
  +0.6%     +27 src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc                     +27  +0.6%
      [NEW]    +782 alts_tsi_handshaker_create                                             +782  [NEW]
  +0.7%     +16 src/core/lib/security/security_connector/alts_security_connector.cc     +16  +0.7%
      +5.0%      +8 alts_server_add_handshakers                                              +8  +5.0%
      +4.8%      +8 alts_channel_add_handshakers                                             +8  +4.8%

  +0.0%    +592 TOTAL                                                               +19.5Ki  +0.2%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                           FILE SIZE
 ++++++++++++++ GROWING                                                             ++++++++++++++
  +0.1%    +393 [None]                                                              +19.4Ki  +0.2%
      +0.1%    +425 [Unmapped]                                                          +19.4Ki  +0.2%
      +6.2%      +8 g_alts_resource                                                           0  [ = ]
   +28%    +156 src/core/tsi/alts_transport_security.cc                                +156   +28%
      [NEW]    +279 grpc_tsi_alts_shared_resource_unref                                    +279  [NEW]
      [NEW]     +52 grpc_tsi_alts_shared_resource_ref                                       +52  [NEW]
       +79%     +19 [Unmapped]                                                              +19   +79%
       +23%     +14 grpc_tsi_alts_init                                                      +14   +23%
      [NEW]      +8 grpc_tsi_alts_shared_resource_get_mu                                     +8  [NEW]
  +1.3%     +59 src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc                     +59  +1.3%
      [NEW]    +805 alts_tsi_handshaker_create                                             +805  [NEW]
      +9.2%      +8 [Unmapped]                                                               +8  +9.2%
  +0.7%     +16 src/core/lib/security/security_connector/alts_security_connector.cc     +16  +0.7%
      +5.0%      +8 alts_server_add_handshakers                                              +8  +5.0%
      +4.8%      +8 alts_channel_add_handshakers                                             +8  +4.8%

  +0.0%    +624 TOTAL                                                               +19.6Ki  +0.2%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,194      Total (>)      1,961,099

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,884,026      Total (>)     10,883,607

 No significant differences in binary sizes


```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,194      Total (>)      1,961,099

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,884,028      Total (>)     10,883,617

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

markdroth commented on src/core/tsi/alts_transport_security.cc:
```diff
@@ -24,19 +24,15 @@
-  return &g_alts_resource;
-}
-
 static void grpc_tsi_alts_wait_for_cq_drain() {
```
If this now requires holding `g_alts_resource.mu`, how about adding a `_locked` suffix to the name?

markdroth commented on src/core/tsi/alts_transport_security.h:
```diff
@@ -33,15 +33,20 @@ typedef struct alts_shared_resource {
+gpr_mu* grpc_tsi_alts_shared_resource_get_mu();
+
+/** This method ref's alts_shared_resource object. */
+alts_shared_resource* grpc_tsi_alts_shared_resource_ref();
```
The suffix `_ref()` seems a bit odd for a function that returns a new instance not previously held by the caller, rather than adding a new ref to an instance that the caller has passed in.  Maybe use `_get()` instead?

Actually, it looks like we expect any given caller to always take one ref, use it for everything they need, and they unref it; there don't seem to be any cases where a given caller will hold more than one ref.  That being the case, how about just calling this `_create()` and changing `_unref()` to `_destroy()`?  That way, the fact that we're actually ref-counting the same object under the hood is abstracted away from the caller.

markdroth commented on src/core/tsi/alts_transport_security.h:
```diff
@@ -33,15 +33,20 @@ typedef struct alts_shared_resource {
- *    TSI handshakes. */
-alts_shared_resource* alts_get_shared_resource(void);
+/** This method returns mu of alts_shared_resource objet. */
+gpr_mu* grpc_tsi_alts_shared_resource_get_mu();
```
This accessor method shouldn't be necessary, since the definition of the `alts_shared_resource` struct is externally visible.  Callers can just directly access its `mu` member.

If you do actually want to make this struct more of an opaque abstraction, then a better approach would be to move the struct definition into the .cc file (leaving only the typedef here) and add methods that allow the necessary actions to be performed without accessing the internal fields directly.  For example, the `init_shared_resources()` function in alts_tsi_handshaker.cc could be replaced with a method called `grpc_tsi_alts_shared_resource_connect()`.

yihuazhang replied:
I agree the accessor method is not necessary. It is removed from the API.

markdroth commented on src/core/tsi/alts/handshaker/alts_tsi_handshaker.h:
```diff
@@ -45,6 +45,8 @@ typedef struct alts_tsi_handshaker alts_tsi_handshaker;
  *
  * - options: ALTS credentials options containing information passed from TSI
  *   caller (e.g., rpc protocol versions).
+ * - resource: an alts_shared_resource object that is shared by all ALTS TSI
```
Please note that the caller is responsible for ensuring that the resources continue to exist until after the handshaker has been destroyed.

yihuazhang replied:
Done.

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                           FILE SIZE
 ++++++++++++++ GROWING                                                             ++++++++++++++
  +0.0%     +33 [None]                                                              +18.7Ki  +0.2%
      +0.0%     +33 [Unmapped]                                                          +18.7Ki  +0.2%
      +6.2%      +8 g_alts_resource                                                           0  [ = ]
   +20%    +108 src/core/tsi/alts_transport_security.cc                                +108   +20%
      [NEW]    +231 grpc_tsi_alts_shared_resource_destroy                                  +231  [NEW]
      [NEW]     +52 grpc_tsi_alts_shared_resource_create                                    +52  [NEW]
       +23%     +14 grpc_tsi_alts_init                                                      +14   +23%
       +29%      +7 [Unmapped]                                                               +7   +29%
  +0.7%     +16 src/core/lib/security/security_connector/alts_security_connector.cc     +16  +0.7%
      +5.0%      +8 alts_server_add_handshakers                                              +8  +5.0%
      +4.8%      +8 alts_channel_add_handshakers                                             +8  +4.8%

 -------------- SHRINKING                                                           --------------
  -0.1%      -5 src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc                      -5  -0.1%
      [DEL]    -733 alts_tsi_handshaker_create                                             -733  [DEL]
      -3.5%     -48 alts_tsi_handshaker_handle_response                                     -48  -3.5%
      [DEL]     -21 _GLOBAL__sub_I_alts_tsi_handshaker.cc                                   -21  [DEL]
      -3.0%      -4 thread_worker                                                            -4  -3.0%

  +0.0%    +152 TOTAL                                                               +18.9Ki  +0.2%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

yihuazhang commented:
All comments are addressed. Besides, I further made the following changes. 

1) In `transport_security.cc`, I moved `self->handshake_shutdown = true` to a correct position. Without this change, `handshaker_shutdown()` in alts_tsi_handshaker.cc will not be correctly executed. 
2) I also added a wait/signal mechanism in `alts_transport_security.cc` to make sure `grpc_tsi_alts_shutdown()` will not be triggered before `grpc_tsi_alts_shared_resource_destroy` correctly returns. It is necessary because when `security_handshaker_shutdown()` is triggered, It is the TSI thread that will execute `grpc_tsi_alts_shared_resource_destroy` API, and since gRPC does not have know about the thread, it may go ahead proceed with `grpc_tsi_alts_shutdown()` while the shared resource has not been destroyed yet.

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,424      Total (>)      1,961,233

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,884,250      Total (>)     10,884,243

 No significant differences in binary sizes


```

yihuazhang commented:
Please hold on reviewing. The test failure still occurs.

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

markdroth commented on src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc:
```diff
@@ -320,7 +320,6 @@ static void thread_worker(void* arg) {
     GPR_ASSERT(event.type != GRPC_QUEUE_TIMEOUT);
     if (event.type == GRPC_QUEUE_SHUTDOWN) {
       /* signal alts_tsi_shutdown() to destroy completion queue. */
```
This comment is stale now.

markdroth commented on src/core/tsi/alts_transport_security.cc:
```diff
@@ -24,53 +24,52 @@
-void grpc_tsi_alts_signal_for_cq_destroy() {
-  gpr_mu_lock(&g_alts_resource.mu);
-  g_alts_resource.is_cq_drained = true;
+static void signal_for_resource_destroy_locked() {
```
s/for_//

markdroth commented on src/core/tsi/alts_transport_security.cc:
```diff
@@ -24,53 +24,52 @@
     grpc_channel_destroy(resource->channel);
     resource->thread.Join();
   }
+  signal_for_resource_destroy_locked();
```
Shouldn't this go inside of the block above, so that it only executes when the refcount goes to zero?

yihuazhang replied:
There could be a situation when refcount is zero but resource->cq is null (e.g., only credential but no channel), and I think we need to take into account that case as well.

yihuazhang replied:
Ah I understand what you mean now. The function should be placed inside the `if` branch. Thanks for pointing it out.

markdroth commented on src/core/tsi/alts_transport_security.cc:
```diff
@@ -24,53 +24,52 @@
-static void grpc_tsi_alts_wait_for_cq_drain() {
-  while (!g_alts_resource.is_cq_drained) {
+static void wait_for_resource_destroy() {
+  gpr_mu_lock(&g_alts_resource.mu);
```
Don't we need to lock and unlock inside of the loop below?  Otherwise, if we're holding the lock when `g_alts_resource.is_resource_destroyed` is false, `grpc_tsi_alts_shared_resource_destroy()` will never be able to get the lock, and therefore it will never be able to set `g_alts_resource.is_resource_destroyed` to true, so we'll never exit this loop.

Also, it seems odd to use both a lock and a cv for shutdown.  Isn't that redundant?

yihuazhang replied:
Good point. I think the lock is not needed.

markdroth commented on src/core/tsi/transport_security.cc:
```diff
@@ -213,10 +213,10 @@ tsi_result tsi_handshaker_next(
   if (self->vtable->shutdown != nullptr) {
     self->vtable->shutdown(self);
   }
+  self->handshake_shutdown = true;
```
Why was this change needed?  What didn't work right in the previous position?

(I'm not objecting to this, just trying to understand.)

yihuazhang replied:
Note that in `handshaker_shutdown()` of alts_tsi_handshaker.cc, since the flag `handshaker_shutdown` has already been set to `true` in the previous position, it will return directly without calling ` alts_handshaker_client_shutdown`.

yihuazhang commented on src/core/tsi/alts_transport_security.cc:
```diff
@@ -24,53 +24,52 @@
-  gpr_mu_lock(&g_alts_resource.mu);
-  g_alts_resource.is_cq_drained = true;
+static void signal_for_resource_destroy_locked() {
+  g_alts_resource.is_resource_destroyed = true;
```
Ah good point. I will unset it in `grpc_tsi_alts_shared_resource_create`.

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                           FILE SIZE
 ++++++++++++++ GROWING                                                             ++++++++++++++
  +0.0%     +33 [None]                                                              +18.8Ki  +0.2%
      +0.0%     +33 [Unmapped]                                                          +18.8Ki  +0.2%
      +6.2%      +8 g_alts_resource                                                           0  [ = ]
   +14%     +76 src/core/tsi/alts_transport_security.cc                                 +76   +14%
      [NEW]    +231 grpc_tsi_alts_shared_resource_destroy                                  +231  [NEW]
      [NEW]     +59 grpc_tsi_alts_shared_resource_create                                    +59  [NEW]
       +23%     +14 grpc_tsi_alts_init                                                      +14   +23%
  +0.7%     +16 src/core/lib/security/security_connector/alts_security_connector.cc     +16  +0.7%
      +5.0%      +8 alts_server_add_handshakers                                              +8  +5.0%
      +4.8%      +8 alts_channel_add_handshakers                                             +8  +4.8%

 -------------- SHRINKING                                                           --------------
  -0.1%      -5 src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc                      -5  -0.1%
      [DEL]    -733 alts_tsi_handshaker_create                                             -733  [DEL]
      -3.5%     -48 alts_tsi_handshaker_handle_response                                     -48  -3.5%
      [DEL]     -21 _GLOBAL__sub_I_alts_tsi_handshaker.cc                                   -21  [DEL]
      -3.0%      -4 thread_worker                                                            -4  -3.0%

  +0.0%    +120 TOTAL                                                               +18.8Ki  +0.2%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,528      Total (>)      1,961,353

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,884,254      Total (>)     10,884,241

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                    FILE SIZE
 ++++++++++++++ GROWING                                      ++++++++++++++
  +0.0%    +352 [None]                                       +9.02Ki  +0.1%
      +0.0%    +288 [Unmapped]                                   +9.02Ki  +0.1%
       +44%     +56 g_alts_resource                                    0  [ = ]
      +0.8%      +8 [None]                                             0  [ = ]
   +48%    +264 src/core/tsi/alts_transport_security.cc         +264   +48%
       +30%     +91 grpc_tsi_alts_shutdown                           +91   +30%
      [NEW]     +67 grpc_tsi_alts_shared_resource_remove_ref         +67  [NEW]
      [NEW]     +51 grpc_tsi_alts_shared_resource_add_ref            +51  [NEW]
      +121%     +29 [Unmapped]                                       +29  +121%
       +43%     +26 grpc_tsi_alts_init                               +26   +43%

  +0.0%    +616 TOTAL                                        +9.28Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,538      Total (>)      1,961,358

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,884,726      Total (>)     10,884,296

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                    FILE SIZE
 ++++++++++++++ GROWING                                      ++++++++++++++
  +0.0%    +352 [None]                                       +9.02Ki  +0.1%
      +0.0%    +288 [Unmapped]                                   +9.02Ki  +0.1%
       +44%     +56 g_alts_resource                                    0  [ = ]
      +0.8%      +8 [None]                                             0  [ = ]
   +48%    +264 src/core/tsi/alts_transport_security.cc         +264   +48%
       +30%     +91 grpc_tsi_alts_shutdown                           +91   +30%
      [NEW]     +67 grpc_tsi_alts_shared_resource_remove_ref         +67  [NEW]
      [NEW]     +51 grpc_tsi_alts_shared_resource_add_ref            +51  [NEW]
      +121%     +29 [Unmapped]                                       +29  +121%
       +43%     +26 grpc_tsi_alts_init                               +26   +43%

  +0.0%    +616 TOTAL                                        +9.28Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,538      Total (>)      1,961,358

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,884,730      Total (>)     10,884,299

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,961,473      Total (>)      1,961,325

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,884,084      Total (>)     10,883,702

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

markdroth commented on src/core/tsi/alts_transport_security.cc:
```diff
@@ -28,38 +28,66 @@ alts_shared_resource* alts_get_shared_resource(void) {
-  while (!g_alts_resource.is_cq_drained) {
-    gpr_cv_wait(&g_alts_resource.cv, &g_alts_resource.mu,
+static void wait_for_cq_drain() {
+  while (!g_alts_resource.is_cq_drained)
```
Please put braces around the body of this loop.

yihuazhang replied:
Done.

markdroth commented on src/core/tsi/alts_transport_security.cc:
```diff
@@ -28,38 +28,66 @@ alts_shared_resource* alts_get_shared_resource(void) {
+}
+
+static void wait_for_resource_destroy() {
+  while (!g_alts_resource.can_destroy_resource)
```
Please put braces around the body of this loop.

yihuazhang replied:
Done

markdroth commented on src/core/tsi/alts_transport_security.cc:
```diff
@@ -28,38 +28,66 @@ alts_shared_resource* alts_get_shared_resource(void) {
-  while (!g_alts_resource.is_cq_drained) {
-    gpr_cv_wait(&g_alts_resource.cv, &g_alts_resource.mu,
+static void wait_for_cq_drain() {
+  while (!g_alts_resource.is_cq_drained)
```
Why is it safe to access `is_cq_drained` without a lock?  And why is it needed in the first place if we're using a condition variable for this?

yihuazhang replied:
I agree that the lock is needed. Without it, the signal could get lost and the thread will indefinitely wait.

markdroth commented on src/core/tsi/alts_transport_security.cc:
```diff
@@ -28,38 +28,66 @@ alts_shared_resource* alts_get_shared_resource(void) {
+}
+
+static void wait_for_resource_destroy() {
+  while (!g_alts_resource.can_destroy_resource)
```
Why is it safe to access `can_destroy_resource` without a lock? And why is it needed in the first place if we're using a condition variable for this?

yihuazhang replied:
Same as above.

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                               FILE SIZE
 ++++++++++++++ GROWING                                 ++++++++++++++
  +0.1%    +416 [None]                                  +8.99Ki  +0.1%
      +0.0%    +320 [Unmapped]                              +8.99Ki  +0.1%
       +44%     +56 g_alts_resource                               0  [ = ]
      +3.9%     +40 [None]                                        0  [ = ]
   +48%    +264 src/core/tsi/alts_transport_security.cc    +264   +48%
       +30%     +91 grpc_tsi_alts_shutdown                      +91   +30%
      [NEW]     +67 grpc_tsi_g_alts_resource_unref              +67  [NEW]
      [NEW]     +51 grpc_tsi_g_alts_resource_ref                +51  [NEW]
       +54%     +33 grpc_tsi_alts_init                          +33   +54%
       +92%     +22 [Unmapped]                                  +22   +92%

  +0.0%    +680 TOTAL                                   +9.25Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,962,848      Total (>)      1,962,660

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,884,083      Total (>)     10,883,696

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                               FILE SIZE
 ++++++++++++++ GROWING                                 ++++++++++++++
  +0.1%    +476 [None]                                  +9.00Ki  +0.1%
      +0.1%    +380 [Unmapped]                              +9.00Ki  +0.1%
       +44%     +56 g_alts_resource                               0  [ = ]
      +3.9%     +40 [None]                                        0  [ = ]
   +46%    +252 src/core/tsi/alts_transport_security.cc    +252   +46%
      [NEW]    +279 grpc_tsi_g_alts_resource_unref             +279  [NEW]
      [NEW]     +51 grpc_tsi_g_alts_resource_ref                +51  [NEW]
       +54%     +33 grpc_tsi_alts_init                          +33   +54%
       +58%     +14 [Unmapped]                                  +14   +58%

  +0.0%    +728 TOTAL                                   +9.25Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,962,923      Total (>)      1,962,660

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,884,088      Total (>)     10,883,701

 No significant differences in binary sizes


```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

yihuazhang commented:
Since we decided to directly apply gRPC thread model (#16773) to resolve the issue, the PR is no longer relevant any more. Thanks Mark for the detailed review.
